File: data-explorer/pipeline/scan.py
"""Wide pitch scan of full LibriSpeech subsets, indexing every scanned sample but
keeping audio ONLY for pitch-crossover voices (female F0 < 180 Hz, male F0 > 140).

Streams each tarball over HTTP and decodes flac bytes in memory — so we never
store the ~60 GB of train audio on disk; only the crossover clips are written.

Usage:
    python -m pipeline.scan train-clean-100 [cap_per_speaker]
    python -m pipeline.scan train-clean-360
    python -m pipeline.scan train-other-500
"""
import io
import os
import sys
import tarfile

import requests
import soundfile as sf

from .datasets import librispeech
from .db import connect, insert_sample, DATA_DIR
from .features import pitch_array

KEEP_DIR = os.path.join(DATA_DIR, "kept")
SPEAKERS = os.path.join(DATA_DIR, "raw", "LibriSpeech", "SPEAKERS.TXT")

FEMALE_MAX = 180.0
MALE_MIN = 140.0


def keep_audio(gender, f0) -> bool:
    if f0 is None:
        return False
    return (gender == "female" and f0 < FEMALE_MAX) or (gender == "male" and f0 > MALE_MIN)


def gender_map() -> dict:
    if not os.path.exists(SPEAKERS):
        raise SystemExit("SPEAKERS.TXT missing — run a small librispeech ingest first "
                         "(e.g. `python -m pipeline.ingest librispeech --subset dev-clean`).")
    out = {}
    with open(SPEAKERS) as f:
        for line in f:
            if line.startswith(";"):
                continue
            p = [x.strip() for x in line.split("|")]
            if len(p) >= 2 and p[0].isdigit():
                out[p[0]] = {"F": "female", "M": "male"}.get(p[1], p[1])
    return out


def scan(subset: str, cap_per_speaker: int = 12):
    con = connect()
    gmap = gender_map()
    os.makedirs(KEEP_DIR, exist_ok=True)
    url = librispeech.SUBSETS[subset]
    print(f"Streaming {subset} …", flush=True)

    per_spk = {}
    n = kept = 0
    with requests.get(url, stream=True, timeout=120) as r:
        r.raise_for_status()
        r.raw.decode_content = False
        with tarfile.open(fileobj=r.raw, mode="r|gz") as tar:
            for m in tar:
                if not m.name.endswith(".flac"):
                    continue
                parts = m.name.split("/")
                if len(parts) < 4:
                    continue
                spk = parts[2]
                if per_spk.get(spk, 0) >= cap_per_speaker:
                    continue
                fh = tar.extractfile(m)
                if fh is None:
                    continue
                try:
                    arr, sr = sf.read(io.BytesIO(fh.read()))
                    feat = pitch_array(arr, sr)
                except Exception:  # noqa: BLE001
                    continue
                per_spk[spk] = per_spk.get(spk, 0) + 1
                gender = gmap.get(spk)
                keep = keep_audio(gender, feat["f0_median"])
                if keep:
                    dst = os.path.join(KEEP_DIR, f"ls-{spk}-{os.path.basename(m.name)}")
                    sf.write(dst, arr, sr, format="FLAC")
                    path = os.path.abspath(dst)
                    kept += 1
                else:
                    path = f"discarded://librispeech/{os.path.basename(m.name)}"
                insert_sample(con, dataset="librispeech", speaker=f"ls-{spk}", gender=gender,
                              path=path, kept=1 if keep else 0, **feat)
                n += 1
                if n % 1000 == 0:
                    con.commit()
                    print(f"  {n} scanned · {kept} kept · {len(per_spk)} speakers", flush=True)
    con.commit()
    con.close()
    print(f"Done {subset}: {n} samples indexed, {kept} kept ({len(per_spk)} speakers).", flush=True)


if __name__ == "__main__":
    sub = sys.argv[1] if len(sys.argv) > 1 else "train-clean-100"
    cap = int(sys.argv[2]) if len(sys.argv) > 2 else 12
    scan(sub, cap)

File: data-explorer/rust/crates/pipeline/src/bin/scan.rs
//! Wide pitch scan of LibriSpeech subsets — Rust port of pipeline/scan.py.
//!
//! Streams a subset tarball over HTTP, decodes each flac in memory, computes
//! pitch-only features, and indexes every sample in samples.db. Audio is kept on
//! disk ONLY for pitch-crossover voices (female F0 < 180, male F0 > 140) — for
//! those the original flac bytes are written verbatim (no re-encode).
//!
//!   scan <subset> <data_dir> [cap_per_speaker]
//! e.g. scan train-clean-100 ../data 12

use std::collections::HashMap;
use std::io::Read;
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use flate2::read::GzDecoder;
use rusqlite::Connection;
use voice_analysis::{audio, pitch_features};

const FEMALE_MAX: f64 = 180.0;
const MALE_MIN: f64 = 140.0;

fn url_for(subset: &str) -> Option<&'static str> {
    Some(match subset {
        "dev-clean" => "https://www.openslr.org/resources/12/dev-clean.tar.gz",
        "test-clean" => "https://www.openslr.org/resources/12/test-clean.tar.gz",
        "dev-other" => "https://www.openslr.org/resources/12/dev-other.tar.gz",
        "test-other" => "https://www.openslr.org/resources/12/test-other.tar.gz",
        "train-clean-100" => "https://www.openslr.org/resources/12/train-clean-100.tar.gz",
        "train-clean-360" => "https://www.openslr.org/resources/12/train-clean-360.tar.gz",
        "train-other-500" => "https://www.openslr.org/resources/12/train-other-500.tar.gz",
        _ => return None,
    })
}

fn keep_audio(gender: Option<&str>, f0: Option<f64>) -> bool {
    match (gender, f0) {
        (Some("female"), Some(f)) => f < FEMALE_MAX,
        (Some("male"), Some(f)) => f > MALE_MIN,
        _ => false,
    }
}

fn main() -> Result<()> {
    let a: Vec<String> = std::env::args().skip(1).collect();
    if a.len() < 2 {
        bail!("usage: scan <subset> <data_dir> [cap_per_speaker]");
    }
    let subset = &a[0];
    let data_dir = PathBuf::from(&a[1]);
    let cap: usize = a.get(2).and_then(|s| s.parse().ok()).unwrap_or(12);
    let url = url_for(subset).ok_or_else(|| anyhow::anyhow!("unknown subset {subset}"))?;

    let kept_dir = data_dir.join("kept");
    std::fs::create_dir_all(&kept_dir)?;
    let speakers = data_dir.join("raw/LibriSpeech/SPEAKERS.TXT");
    if !speakers.exists() {
        bail!(
            "SPEAKERS.TXT missing at {} — ingest a small subset first",
            speakers.display()
        );
    }
    let gmap = gender_map(&speakers)?;
    let mut con = Connection::open(data_dir.join("samples.db"))?;
    con.execute_batch(SCHEMA)?;
    con.execute_batch("PRAGMA busy_timeout=60000;")?;

    eprintln!("Streaming {subset} …");
    let resp = ureq::get(url).call().context("http get")?;
    let gz = GzDecoder::new(resp.into_reader());
    let mut ar = tar::Archive::new(gz);

    let mut per_spk: HashMap<String, usize> = HashMap::new();
    let (mut n, mut kept) = (0usize, 0usize);
    let tx = con.transaction()?;
    for entry in ar.entries()? {
        let mut entry = entry?;
        let name = entry.path()?.to_string_lossy().into_owned();
        if !name.ends_with(".flac") {
            continue;
        }
        let parts: Vec<&str> = name.split('/').collect();
        if parts.len() < 4 {
            continue;
        }
        let spk = parts[2].to_string();
        if per_spk.get(&spk).copied().unwrap_or(0) >= cap {
            continue;
        }
        let mut bytes = Vec::new();
        entry.read_to_end(&mut bytes)?;
        let (samples, sr) = match audio::decode_flac_bytes(&bytes) {
            Ok(v) => v,
            Err(_) => continue,
        };
        let feat = pitch_features(&samples, sr);
        *per_spk.entry(spk.clone()).or_insert(0) += 1;
        let gender = gmap.get(&spk).map(|s| s.as_str());
        let base = Path::new(&name).file_name().unwrap().to_string_lossy();
        let keep = keep_audio(gender, feat.f0_median);
        let path = if keep {
            let dst = kept_dir.join(format!("ls-{spk}-{base}"));
            std::fs::write(&dst, &bytes)?; // original flac bytes, verbatim
            kept += 1;
            std::fs::canonicalize(&dst)?.to_string_lossy().into_owned()
        } else {
            format!("discarded://librispeech/{base}")
        };
        tx.execute(
            "INSERT OR IGNORE INTO samples
             (dataset,speaker,gender,path,duration,f0_median,voiced_frac,f1_median,f2_median,kept)
             VALUES ('librispeech',?,?,?,?,?,?,NULL,NULL,?)",
            rusqlite::params![
                format!("ls-{spk}"),
                gender,
                path,
                feat.duration,
                feat.f0_median,
                feat.voiced_frac,
                keep as i64
            ],
        )?;
        n += 1;
        if n % 1000 == 0 {
            eprintln!("  {n} scanned · {kept} kept · {} speakers", per_spk.len());
        }
    }
    tx.commit()?;
    eprintln!(
        "Done {subset}: {n} samples indexed, {kept} kept ({} speakers).",
        per_spk.len()
    );
    Ok(())
}

fn gender_map(path: &Path) -> Result<HashMap<String, String>> {
    let text = std::fs::read_to_string(path)?;
    let mut out = HashMap::new();
    for line in text.lines() {
        if line.starts_with(';') {
            continue;
        }
        let p: Vec<&str> = line.split('|').map(|s| s.trim()).collect();
        if p.len() >= 2 && !p[0].is_empty() && p[0].chars().all(|c| c.is_ascii_digit()) {
            let g = match p[1] {
                "F" => "female",
                "M" => "male",
                other => other,
            };
            out.insert(p[0].to_string(), g.to_string());
        }
    }
    Ok(out)
}

const SCHEMA: &str = "
CREATE TABLE IF NOT EXISTS samples (
    id integer primary key autoincrement,
    dataset text not null, speaker text not null, gender text, path text not null,
    duration real, f0_median real, voiced_frac real, f1_median real, f2_median real,
    kept integer default 1,
    unique(dataset, path));
";
